Unescape does not handle the two byte sequence \n

A JSON text containing the two characters \n (bytes 0x5C and 0x6E) is not translated to a new line (byte 0x0A).

This is explained for example in http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf in page 4: "There are two-character escape sequence representations of some characters." with the following list: \", \\, \/, \b, \f, \n, \r and \t

Of course. I also realises that I didn't explain what's actually happening: It seems that GetString converts two byte sequence \n into just a single byte n instead of a single byte new-line

    func TestENDL(t *testing.T) {
	  s, e := jsonparser.GetString([]byte("{\"multiline\":\"line 1\\nline 2\"}"), "multiline")
	  assert.Equal(t, string(s), "line 1\nline2") // Not equal: "line 1nline 2" != "line 1\nline2"
    }

That's the test case which I used to pinpoint the problem. I also added (what I think is a correct test case) test case `{in: "\\n", out: "\n", canAlloc: true},`into the escape_test.go into the unescapeTest{} structure (as in this diff):

    vagrant@precise64:/Development/jsonparser$ git diff
    diff --git a/escape_test.go b/escape_test.go
    index db18108..0fb5d22 100644
    --- a/escape_test.go
    +++ b/escape_test.go
    @@ -125,6 +125,7 @@ var unescapeTests = []unescapeTest{
            {in: `abcde\uD800`, isErr: true},
            {in: `ab\uD800de`, isErr: true},
            {in: `\uD800abcde`, isErr: true},
    +        {in: "\\n", out: "\n", canAlloc: true},
     }

which fails in the following way:

    === RUN   TestUnescape
    --- FAIL: TestUnescape (0.00s)
    	escape_test.go:183: Unescape(`\n`, bufsize=0) returned unescaped mismatch: expected `
    		` ([10], len 1), obtained `n` ([110], len 1)

What do you think? I didn't test any other items in the list than \n

--- a/escape.go
+++ b/escape.go
@@ -69,6 +69,18 @@ func decodeUnicodeEscape(in []byte) (rune, int) {
 
 }
 
+// backslashCharEscapeTable: when '\X' is found for some byte X, it is to be replaced with backslashCharEscapeTable[X]
+var backslashCharEscapeTable = [...]byte{
+	'"':  '"',
+	'\\': '\\',
+	'/':  '/',
+	'b':  '\b',
+	'f':  '\f',
+	'n':  '\n',
+	'r':  '\r',
+	't':  '\t',
+}
+
 // unescapeToUTF8 unescapes the single escape sequence starting at 'in' into 'out' and returns
 // how many characters were consumed from 'in' and emitted into 'out'.
 // If a valid escape sequence does not appear as a prefix of 'in', (-1, -1) to signal the error.
@@ -80,9 +92,9 @@ func unescapeToUTF8(in, out []byte) (inLen int, outLen int) {
 
 	// https://tools.ietf.org/html/rfc7159#section-7
 	switch e := in[1]; e {
-	case '"', '\\', 'n', 't', 'r', '/', 'b', 'f':
-		// Valid basic 2-character escapes
-		out[0] = e
+	case '"', '\\', '/', 'b', 'f', 'n', 'r', 't':
+		// Valid basic 2-character escapes (use lookup table)
+		out[0] = backslashCharEscapeTable[e]
 		return 2, 1
 	case 'u':
 		// Unicode escape

--- a/parser_test.go
+++ b/parser_test.go
@@ -168,6 +168,13 @@ var getTests = []GetTest{
 		data:    `1`,
 	},
 	GetTest{
+		desc:    `key and value with whitespace escapes`,
+		json:    `{"key\b\f\n\r\tkey":"value\b\f\n\r\tvalue"}`,
+		path:    []string{"key\b\f\n\r\tkey"},
+		isFound: true,
+		data:    `value\b\f\n\r\tvalue`, // value is not unescaped since this is Get(), but the key should work correctly
+	},
+	GetTest{
 		desc:    `key with Unicode escape`,
 		json:    `{"a\u00B0b":1}`,
 		path:    []string{"a\u00B0b"},
@@ -370,6 +377,13 @@ var getStringTests = []GetTest{
 		path:    []string{"c"},
 		isFound: true,
 		data:    `\"`,
+	},
+	GetTest{
+		desc:    `key and value with whitespace escapes`,
+		json:    `{"key\b\f\n\r\tkey":"value\b\f\n\r\tvalue"}`,
+		path:    []string{"key\b\f\n\r\tkey"},
+		isFound: true,
+		data:    "value\b\f\n\r\tvalue", // value is unescaped since this is GetString()
 	},
 }
 

